### PR TITLE
Optimize Softmax

### PR DESCRIPTION
**Pull request by daemyung**
## 🙏 Describe the pull request

A backward of Softmax is implemented into a kernel.

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: trident/language/language.py [hunk 1]
```diff
@@ -60,6 +71,11 @@ def make_group_msk(blk, grp_sz, off, h):
     return blk[:, None] & (triton.language.arange(0, grp_sz) + off < h)[None, :]
 
 
+@triton.jit
+def matmul(a, b):
+    return triton.language.sum(a * triton.language.trans(b), 1)
```
Comment (kakao-steve-ai) [inline trident/language/language.py]: 일반적인 행렬곱셈 함수가 아니라, 벡터 행렬 곱셈에서만 유효한거 같아서요. 함수이름을 matmul 대신 vec_mat_mul 같은걸로 하면 어떨까요?

Comment (daemyung) [reply trident/language/language.py]: 좋은 커멘트 감사합니다. 일반적으로 사용되는 `gemv`로 변경하였습니다.

File: tests/test_softmax.py [hunk 2]
```diff
@@ -34,9 +34,9 @@ def test_forward(num_vec, vec_sz, dtype, device):
 
 
 @pytest.mark.parametrize("num_vec, vec_sz", [(4, 64), (3, 128)])
```
Comment (kakao-steve-ai) [inline tests/test_softmax.py]: vec_sz가 16의 배수가 아닌 경우가 테스트에 추가되면 좋겠습니다. 로컬에서 (5, 70) 추가해서 돌려봤는데 에러가 발생하네요. root cause는 좀 더 확인해봐야겠습니다.

Comment (daemyung) [reply tests/test_softmax.py]: 반드시 2의 거듭제곱이어야 하는데 16의 배수가 2의 거듭제곱이 아니여서 발생하는 문제였습니다. 수정하였습니다.

File: tests/test_softmax.py [hunk 3]
```diff
@@ -33,10 +33,10 @@ def test_forward(num_vec, vec_sz, dtype, device):
     assert util.equal(torch.nn.Softmax(1).forward(inp), trident.Softmax(1).forward(inp))
 
 
-@pytest.mark.parametrize("num_vec, vec_sz", [(4, 64), (3, 128)])
-def test_backward(num_vec, vec_sz, dtype, device):
-    inp = torch.randn(num_vec, vec_sz, dtype=dtype, device=device)
-    tgt = torch.randn(num_vec, vec_sz, dtype=dtype, device=device)
+@pytest.mark.parametrize("num_vec, vec_sz", [(4, 64), (5, 70)])
```
Comment (daemyung) [inline tests/test_softmax.py:36]: @rbbang 실패하는 테스트 케이스로 변경하였습니다.

**Review by kakao-steve-ai — APPROVED** [review]
LGTM
